### PR TITLE
refactor layer_norm and expose out variant

### PR DESCRIPTION
**Pull request by makslevental**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61824
* __->__ #61821
* #61802
* #61801
* #61800

Differential Revision: [D29812516](https://our.internmc.facebook.com/intern/diff/D29812516)

### CODE REVIEW

**Comment by makslevental** [comment]
This PR exposes the out variant of `layer_norm`. As well it refactors because there was a large amount of shared code between `layer_norm.cpp` and `layer_norm_kernel.cu`. 

It took some experimentation to get the codegen and existing calls to cooperate so I am more than open/receptive to better refactoring suggestions.

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/26288

If you're continuing this theme, out/inplace BatchNorm would also be very useful for invertible nets

**Comment by makslevental** [comment]
@vadimkantorov looks you've already added [batch_norm_cuda_out](https://github.com/pytorch/pytorch/pull/29192) (good job!). which backend in particular are you looking for to be added? skimming that issue it looks you're mostly hung up on the backward pass?

**Comment by vadimkantorov** [comment]
I never did the binding work to expose this to Python in public functions as F.batch_norm, never implemented inplace option for _BatchNorm. 

So yes, it's possible, but I end up calling manually a bunch of functions and it's slower than nn.BatchNorm

My impl:
```
class InplaceBatchNorm1d(nn.BatchNorm1d):
	def forward(self, input):
		return InplaceBatchNorm1d.Function.apply(
			input, self.weight, self.bias, self.running_mean, self.running_var, self.eps, self.momentum, self.training
		)

	class Function(torch.autograd.function.Function):
		@staticmethod
		def forward(ctx, input, weight, bias, running_mean, running_var, eps, momentum, training):
			mean, var = torch.batch_norm_update_stats(input, running_mean, running_var, momentum) if training else (running_mean, running_var)
			invstd = (var + eps).rsqrt_()
			output = torch.batch_norm_elemt(input, weight, bias, mean, invstd, 0, out = input)
			ctx.training = training
			ctx.save_for_backward(input, weight, bias, mean, invstd)
			ctx.mark_dirty(input)
			return input

		@staticmethod
		def backward(ctx, grad_output):
			assert ctx.training, 'Inplace BatchNorm supports only training mode, input gradients with running stats used are not yet supported'
			saved_output, weight, bias, mean, invstd = ctx.saved_tensors
			saved_input = torch.batch_norm_elemt(
				saved_output, invstd.reciprocal(), mean, bias, weight.reciprocal(), 0, out = saved_output
			)
			sum_dy, sum_dy_xmu, grad_weight, grad_bias = torch.batch_norm_backward_reduce(grad_output, saved_input, mean, invstd, weight, *ctx.needs_input_grad[:3])
			divisor = saved_input.numel() // saved_input.size(1)
			mean_dy = sum_dy.div_(divisor)
			mean_dy_xmu = sum_dy_xmu.div_(divisor)
			grad_input = torch.batch_norm_backward_elemt(
				grad_output, saved_input, mean, invstd, weight, mean_dy, mean_dy_xmu
			)
			return grad_input, grad_weight, grad_bias, None, None, None, None, None
```

**Comment by vadimkantorov** [comment]
Relevant piece of discussion about invertible nets, if you are interested ... of maybe pushing this direction :)  https://github.com/pytorch/pytorch/issues/23756#issuecomment-792338938

**Comment by makslevental** [comment]
@vadimkantorov out variant is automatically exposed; check `.../python3.9/site-packages/torch/_C/_VariableFunctions.pyi`:

image

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812516).

**Comment by vadimkantorov** [comment]
It would be cool if it could be directly exposed in F.layer_norm and nn.LayerNorm(inplace) and same for other normalization variants

**Comment by vadimkantorov** [comment]
Then probably some hacks to bypass version modification is needed :( Otherwise a way to manually call the backward

**Comment by makslevental** [comment]
@desertfire , @hlu1 , @eellison and @Krovatkin  any blockers here?

**Comment by desertfire** [comment]
> @desertfire , @hlu1 , @eellison and @Krovatkin any blockers here?

Can you give https://pytorch.org/tutorials/recipes/recipes/benchmark.html a try to make sure there is no perf regression to layer_norm, given this is a non-trivial refactoring?

**Comment by makslevental** [comment]
> Can you give https://pytorch.org/tutorials/recipes/recipes/benchmark.html a try to make sure there is no perf regression to layer_norm, given this is a non-trivial refactoring?

@desertfire 

https://gist.github.com/makslevental/8d606368f05e54483a211e32b67949c8

I ran this by flipping between branches (after merging latest master). not sure what to conclude (most times are lower but a few are higher).

**Review by desertfire — APPROVED** [review]

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812516).
